frappe/app.py

```diff
@@ -165,9 +121,62 @@ def make_form_dict(request):
 		# _ is passed by $.ajax so that the request is not cached by the browser. So, remove _ from form_dict
 		frappe.local.form_dict.pop("_")
 
-application = local_manager.make_middleware(application)
-application.debug = True
+def handle_exception(e):
+	http_status_code = getattr(e, "http_status_code", 500)
+
+	if (http_status_code==500
+		and isinstance(e, MySQLdb.OperationalError)
+		and e.args[0] in (1205, 1213)):
+			# 1205 = lock wait timeout
+			# 1213 = deadlock
+			# code 409 represents conflict
+			http_status_code = 508
+
+	if frappe.local.is_ajax or 'application/json' in request.headers.get('Accept', ''):
+		response = frappe.utils.response.report_error(http_status_code)
+	else:
+		traceback = "<pre>"+frappe.get_traceback()+"</pre>"
+		if frappe.local.flags.disable_traceback:
+			traceback = ""
+
+		frappe.respond_as_web_page("Server Error",
+			traceback,
+			http_status_code=http_status_code)
+		response = frappe.website.render.render("message", http_status_code=http_status_code)
+
+	if e.__class__ == frappe.AuthenticationError:
+		if hasattr(frappe.local, "login_manager"):
+			frappe.local.login_manager.clear_cookies()
+
+	if http_status_code==500:
+		logger.error('Request Error')
+
+	make_error_snapshot(e)
 
+	return response
+
+def after_request(rollback):
+	if (frappe.local.request.method in ("POST", "PUT") or frappe.local.flags.commit) and frappe.db:
+		if frappe.db.transaction_writes:
+			frappe.db.commit()
+			rollback = False
+
+	# update session
+	if getattr(frappe.local, "session_obj", None):
+		updated_in_db = frappe.local.session_obj.update()
+		if updated_in_db:
+			frappe.db.commit()
+			rollback = False
+
+	call_hook_method("after_request")
```

Comment:
@rmehta is using hooks overkill and should I have just imported and called the function?